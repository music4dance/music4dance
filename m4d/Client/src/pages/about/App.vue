<template>
  <page id="app" title="music4dance background info" :breadcrumbs="breadcrumbs">
    <h2>About music4dance</h2>
    <p>
      <b>music4dance</b> is a project that has been floating around in my head
      in various forms since I first started ballroom dancing - so what
      <i>is</i> music4dance, really?
    </p>
    <ul>
      <li>
        <b>music4dance</b> is meant to help find music to dance to as well has
        help figure out what dances can be danced to a piece of music
      </li>
      <li>
        <b>music4dance</b> intended to help DJs find new music to add to their
        playlists as well as for students and teachers to build playlists to
        practice to
      </li>
      <li>
        <b>music4dance</b> meant to help dancers understand the relationship
        between dance and the music that they dance to
      </li>
      <li>
        <b>music4dance</b> a place for people to find music to dance to for
        special occassions like weddings and holidy parties
      </li>
    </ul>
    <p>
      Even to make a dent in any one of those goals is pretty ambitious. That's
      about all I've done is to make a dent. I certainly haven't solved any of
      these problems to my satisfaction. So most of all music4dance is a work in
      progress and a playground for me to experiment with the intersection of
      three of my lifelong passions - music, dance and programming.
    </p>
    <p>
      And while I'm doing that, I hope that music4dance provides at least a
      partial solution to some of the above for you. I'm sure that's true for
      some and I can say that some of the most interesting features of this site
      have started as questions from the community. So please keep using the
      site and <a href="https://music4dance.blog/feedback/">let me know</a> what
      you think.
    </p>
    <p class="neutral">
      Please read my <a href="https://www.music4dance.blog">blog</a> for more
      about the history and motivation behind
      <a href="https://www.music4dance.net">music4dance</a>.
    </p>
    <h2>About me</h2>
    <p>
      I’m an amateur dancer and with some training in music who also happens to
      be a professional software engineer. I love dancing, and I particularly
      love dancing when I enjoy the music that I’m dancing to.
    </p>
    <p>
      My dance background is primarily in Ballroom. I’ve been an amateur
      competitor in the
      <a href="https://www.music4dance.net/dances/american-rhythm"
        >American Rhythm style</a
      >). Recently I've moved more towards social partner dancing. As I was
      starting to work on music4dance in earnest, I dove into tap.
    </p>
    <p>
      Most of my programming time outside of my main day jobs has been related
      to music.
    </p>
    <ul>
      <li>
        When the Macintosh came out, my first thought was - Music Editor. And I
        wrote one - although I never got it to the point where I could sell it.
      </li>
      <li>
        My work study project in college was writing music theory software for
        the Macintosh
      </li>
      <li>
        When I left Corporate America to try my hand at the startup world I
        joined a company that was digitizing sheet music and did the initial
        work to take that online.
      </li>
      <li>
        When I left Corporate America for the second time to rediscover my love
        of programming (and generally find myself) I founded music4dance
      </li>
    </ul>
    <h2>My Resume</h2>
    <p>
      <a href="/Content/DavidGray-Resume.pdf">Printable Version</a>
    </p>
    <p>
      <a href="/Home/Resume">Web Version</a>
    </p>
  </page>
</template>

<script lang="ts">
import Page from "@/components/Page.vue";
import { BreadCrumbItem, infoTrail } from "@/model/BreadCrumbItem";
import { Component, Vue } from "vue-property-decorator";

@Component({
  components: {
    Page,
  },
})
export default class App extends Vue {
  private breadcrumbs: BreadCrumbItem[] = [
    ...infoTrail,
    { text: "About", active: true },
  ];
}
</script>
