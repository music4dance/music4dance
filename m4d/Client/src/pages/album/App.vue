<template>
  <page id="app">
    <h1>
      Album: {{ model.title }}
      <span v-if="model.artist"
        >by <a :href="artistRef">{{ model.artist }}</a></span
      >
    </h1>
    <song-table
      :histories="model.histories"
      :filter="model.filter"
      :hideSort="true"
      :hiddenColumns="hidden"
      @song-selected="selectSong"
    ></song-table>
    <admin-footer :model="model" :selected="selected"></admin-footer>
  </page>
</template>

<script lang="ts">
import Page from "@/components/Page.vue";
import SongTable from "@/components/SongTable.vue";
import AdminFooter from "@/components/AdminFooter.vue";
import { safeEnvironment } from "@/helpers/DanceEnvironmentManager";
import { AlbumModel } from "@/model/AlbumModel";
import { DanceEnvironment } from "@/model/DanceEnvironment";
import "reflect-metadata";
import { TypedJSON } from "typedjson";
import SongSelector from "@/mix-ins/SongSelector";

declare const model: string;

export default SongSelector.extend({
  components: { AdminFooter, Page, SongTable },
  data() {
    return new (class {
      model: AlbumModel = TypedJSON.parse(model, AlbumModel)!;
      environment: DanceEnvironment = safeEnvironment();
    })();
  },
  computed: {
    artistRef(): string {
      return `/song/artist?name=${encodeURIComponent(this.model.artist || "")}`;
    },
    hidden(): string[] {
      return this.model.artist ? ["Artist"] : [];
    },
  },
});
</script>
