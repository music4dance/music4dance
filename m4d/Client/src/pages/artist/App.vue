<template>
  <page id="app" :title="title" :consumesEnvironment="true">
    <song-table 
      :songs="model.songs"
      :filter="model.filter"
      :userName="model.userName"
      :hideSort="true"
      :hiddenColumns="['Artist']"
    ></song-table>
  </page>
</template>

<script lang="ts">
// tslint:disable: max-classes-per-file
import 'reflect-metadata';
import { Component, Vue } from 'vue-property-decorator';
import SongTable from '@/components/SongTable.vue';
import Page from '@/components/Page.vue';
import { TypedJSON } from 'typedjson';
import { Song } from '@/model/Song';
import { ArtistModel } from '@/model/ArtistModel';

declare const model: string;

@Component({
  components: {
    Page,
    SongTable,
  },
})
export default class App extends Vue {
  private readonly model: ArtistModel;

  constructor() {
    super();

    this.model = TypedJSON.parse(model, ArtistModel)!;
  }

  private get title(): string {
    return `Artist: ${this.model.artist}`;
  }
}
</script>
