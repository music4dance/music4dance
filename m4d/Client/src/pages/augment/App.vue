<template>
  <page
    id="app"
    :consumesEnvironment="true"
    @environment-loaded="onEnvironmentLoaded"
  >
    <b-row v-if="phase === 'lookup'">
      <b-col>
        <b-alert v-if="lastSong" dismissible show="10">
          Thank you for {{ created ? "adding" : "editing" }}
          <i>{{ lastSong.title }}</i> by {{ lastSong.artist }}
        </b-alert>
        <h1>Add Song</h1>
        <b-tabs v-if="canAugment" v-model="tabIndex" card>
          <b-tab title="by Title"
            ><b-card-text
              ><augment-search
                @edit-song="editSong($event)"
              ></augment-search></b-card-text
          ></b-tab>
          <b-tab title="by Id">
            <b-card-text
              ><augment-lookup @edit-song="editSong($event)"></augment-lookup
            ></b-card-text>
          </b-tab>
          <b-tab v-if="isAdmin" title="Admin">
            <b-card-text>
              <p>Paste in text version of propery list</p>
              <b-input-group prepend="Properties">
                <b-form-input
                  id="admin-properties"
                  palceholder="Song Properties TSV"
                  v-model="propertiesString"
                  trim
                ></b-form-input>
                <b-button variant="primary" @click="adminCreate"
                  >Create</b-button
                >
              </b-input-group>
            </b-card-text>
          </b-tab>
        </b-tabs>
        <augment-info v-else> </augment-info>
      </b-col>
    </b-row>
    <div v-else>
      <b-alert show variant="success">
        <div v-if="songModel.created">
          <b>Create Song:</b> This song is new to music4dance, please fill in
          missing fields and click <b>Add Song</b> to add this song to the
          catalog. Remember, you must vote on at least one dance to add this
          song.
        </div>
        <div v-else>
          <b>Edit Song:</b> We found this song in the music4dance catalog,
          please vote on dance styles and add tags to improve the catalog.
        </div>
      </b-alert>
      <song-core
        :model="songModel"
        :environment="environment"
        :startEditing="true"
        :creating="songModel.created"
        @song-saved="reset(true)"
        @cancel-changes="reset(false)"
      ></song-core>
    </div>
  </page>
</template>

<script lang="ts">
import "reflect-metadata";
import { Component, Mixins } from "vue-property-decorator";
import AdminTools from "@/mix-ins/AdminTools";
import AugmentInfo from "./components/AugmentInfo.vue";
import AugmentLookup from "./components/AugmentLookup.vue";
import AugmentSearch from "./components/AugmentSearch.vue";
import SongCore from "@/pages/song/components/SongCore.vue";
import Page from "@/components/Page.vue";
import { SongDetailsModel } from "@/model/SongDetailsModel";
import { SongFilter } from "@/model/SongFilter";
import { DanceEnvironment } from "@/model/DanceEnvironmet";
import { Song } from "@/model/Song";
import { SongHistory } from "@/model/SongHistory";

enum AugmentPhase {
  lookup = "lookup",
  results = "results",
  edit = "edit",
}

@Component({
  components: { AugmentInfo, AugmentLookup, AugmentSearch, SongCore, Page },
})
export default class App extends Mixins(AdminTools) {
  private phase = AugmentPhase.lookup;
  private songModel: SongDetailsModel | null = null;
  private lastSong: Song | null = null;
  private created = false;
  private environment: DanceEnvironment | null = null;
  private propertiesString = "";
  private tabIndex = 0;

  private get canAugment(): boolean {
    return this.canTag || this.isPremium;
  }

  private editSong(model: SongDetailsModel): void {
    this.songModel = new SongDetailsModel({
      created: !!model.created,
      songHistory: model.songHistory,
      filter: new SongFilter(),
      userName: this.userName,
    });
    this.phase = AugmentPhase.edit;
  }

  private async adminCreate(): Promise<void> {
    console.log(`Create Song: ${this.propertiesString}`);
    this.songModel = new SongDetailsModel({
      created: true,
      songHistory: SongHistory.fromString(this.propertiesString),
      filter: new SongFilter(),
      userName: this.userName,
    });
    this.propertiesString = "";
    this.phase = AugmentPhase.edit;
  }

  private reset(saved: boolean): void {
    if (saved) {
      this.lastSong = Song.fromHistory(this.songModel!.songHistory);
      this.created = !!this.songModel!.created;
    }
    this.phase = AugmentPhase.lookup;
    this.songModel = null;
  }

  private onEnvironmentLoaded(environment: DanceEnvironment): void {
    this.environment = environment;
  }
}
</script>

<style lang="scss" scoped>
#service-id {
  width: 25rem;
}
</style>
