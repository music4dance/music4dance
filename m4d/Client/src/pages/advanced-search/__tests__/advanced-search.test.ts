import { Preloads } from "@/Preloads";
import { testPageAsync } from "@/helpers/TestHelpers";
import App from "../App.vue";

describe("AdvancedSearch.vue", () => {
  const originalWindowLocation = window.location;

  beforeEach(() => {
    Object.defineProperty(window, "location", {
      configurable: true,
      enumerable: true,
      value: new URL(
        "https://localhost:5001/song/advancedsearchform?filter=v2-Advanced-BOL%2CRMB--Love"
      ),
    });
  });

  afterEach(() => {
    Object.defineProperty(window, "location", {
      configurable: true,
      enumerable: true,
      value: originalWindowLocation,
    });
  });

  test("renders the advanced search page", async () => {
    await testPageAsync(App, Preloads.All);
  });
});
