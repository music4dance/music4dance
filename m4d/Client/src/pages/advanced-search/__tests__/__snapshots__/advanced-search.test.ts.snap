// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedSearch.vue renders the advanced search page 1`] = `
<div>
  <div>
    <!---->
    <!---->
    <nav class="navbar navbar-dark bg-primary navbar-expand-lg" id="mainMenu"><a href="/" target="_self" class="navbar-brand"><img src="/images/header-logo.png" height="40" title="music4dance"></a> <button type="button" aria-label="Toggle navigation" class="navbar-toggler collapsed" aria-expanded="false" aria-controls="nav-collapse"><span class="navbar-toggler-icon"></span></button>
      <transition-stub css="true" enterclass="" leaveclass="collapse show" entertoclass="collapse show" leavetoclass="collapse" enteractiveclass="collapsing" leaveactiveclass="collapsing">
        <div id="nav-collapse" class="navbar-collapse collapse" style="display: none;">
          <ul class="navbar-nav">
            <li id="music-menu" class="nav-item b-nav-dropdown dropdown"><a id="music-menu__BV_toggle_" role="button" aria-haspopup="true" aria-expanded="false" aria-controls="music-menu__BV_toggle_menu_" href="#music-menu" target="_self" class="nav-link dropdown-toggle"><span>Music</span></a>
              <ul tabindex="-1" aria-labelledby="music-menu__BV_toggle_" id="music-menu__BV_toggle_menu_" class="dropdown-menu">
                <li role="presentation"><a role="menuitem" href="/dances" target="_self" class="dropdown-item">Dances</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/ballroom-competition-categories" target="_self" class="dropdown-item">Ballroom</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/latin" target="_self" class="dropdown-item">Latin</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/swing" target="_self" class="dropdown-item">Swing</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/tango" target="_self" class="dropdown-item">Tango</a></li>
                <li role="presentation"><a role="menuitem" href="/song" target="_self" class="dropdown-item">Song Library</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/song/advancedsearchform" target="_self" class="dropdown-item">Advanced Search</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/song/augment" target="_self" class="dropdown-item">Add Song</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/song/newmusic" target="_self" class="dropdown-item">New Music</a></li>
                <li role="presentation"><a role="menuitem" href="/dances/wedding-music" target="_self" class="dropdown-item">Wedding</a></li>
                <li role="presentation"><a role="menuitem" href="/song/holidaymusic" target="_self" class="dropdown-item">Holiday</a></li>
                <li role="presentation"><a role="menuitem" href="/song/addtags/?tags=%2BHalloween%3AOther" target="_self" class="dropdown-item">Halloween</a></li>
                <li role="presentation"><a role="menuitem" href="/tag" target="_self" class="dropdown-item">Tags</a></li>
              </ul>
            </li>
            <li id="tools-menu" class="nav-item b-nav-dropdown dropdown"><a id="tools-menu__BV_toggle_" role="button" aria-haspopup="true" aria-expanded="false" aria-controls="tools-menu__BV_toggle_menu_" href="#tools-menu" target="_self" class="nav-link dropdown-toggle"><span>Tools</span></a>
              <ul tabindex="-1" aria-labelledby="tools-menu__BV_toggle_" id="tools-menu__BV_toggle_menu_" class="dropdown-menu">
                <li role="presentation"><a role="menuitem" href="/home/counter" target="_self" class="dropdown-item">Tempo Counter</a></li>
                <li role="presentation"><a role="menuitem" href="/home/tempi" target="_self" class="dropdown-item">Tempi (Tempos)</a></li>
                <li role="presentation"><a role="menuitem" href="/song/advancedsearchform" target="_self" class="dropdown-item">Advanced Search</a></li>
                <!---->
              </ul>
            </li>
            <li class="nav-item b-nav-dropdown dropdown"><a role="button" aria-haspopup="true" aria-expanded="false" href="#" target="_self" class="nav-link dropdown-toggle"><span>Info</span></a>
              <ul tabindex="-1" class="dropdown-menu">
                <li role="presentation"><a role="menuitem" href="https://music4dance.blog/music4dance-help/song-list/" target="_self" class="dropdown-item">Help</a></li>
                <li role="presentation"><a role="menuitem" href="https://music4dance.blog/" target="_self" class="dropdown-item">Blog</a></li>
                <li role="presentation"><a role="menuitem" href="/home/faq" target="_self" class="dropdown-item">FAQ</a></li>
                <li role="presentation"><a role="menuitem" href="/home/about" target="_self" class="dropdown-item">About music4dance</a></li>
                <li role="presentation"><a role="menuitem" href="/home/readinglist" target="_self" class="dropdown-item">Reading List</a></li>
                <li role="presentation"><a role="menuitem" href="/home/sitemap" target="_self" class="dropdown-item">Site Map</a></li>
                <li role="presentation"><a role="menuitem" href="/home/privacypolicy" target="_self" class="dropdown-item">Privacy Policy</a></li>
                <li role="presentation"><a role="menuitem" href="/home/credits" target="_self" class="dropdown-item">Credits</a></li>
              </ul>
            </li>
            <li id="contribute-menu" right="" class="nav-item"><a href="/home/contribute" target="_self" class="nav-link">Contribute</a></li>
            <!---->
          </ul>
          <ul class="navbar-nav ml-auto">
            <li class="form-inline">
              <form class="form-inline"><input type="text" placeholder="Search" class="mr-sm-2 form-control form-control-sm"> <button type="submit" disabled="disabled" class="btn my-2 my-sm-0 btn-secondary btn-sm disabled">Search</button></form>
            </li>
            <li right="" class="nav-item"><a href="/identity/account/register?returnUrl=/song/advancedsearchform?filter=v2-Advanced-BOL%2CRMB--Love" target="_self" class="nav-link">Register</a></li>
            <li right="" class="nav-item"><a href="/identity/account/login?returnUrl=/song/advancedsearchform?filter=v2-Advanced-BOL%2CRMB--Love" target="_self" class="nav-link">Login</a></li>
          </ul>
        </div>
      </transition-stub>
    </nav>
    <!---->
    <form id="logoutForm" action="/identity/account/logout" method="post" style="height: 0px;"><input name="__RequestVerificationToken" type="hidden" value="FOO"> <input type="hidden" name="returnUrl" value="/"> <button id="logout" type="submit" class="btn btn-link"></button></form>
  </div>
  <!---->
  <div id="body-content" class="container-fluid body-content">
    <div class="mt-2"></div>
    <div>
      <h1 class="col-sm-12" style="font-size: 22px; text-align: center;">
        Advanced Song Search
      </h1>
      <div style="max-width: 600px; margin-left: auto; margin-right: auto;">
        <form id="advanced-search" novalidate="novalidate" class="">
          <div id="searchStringGroup" role="group" class="form-group"><label id="searchStringGroup__BV_label_" for="searchString" class="d-block">Keywords:</label>
            <div><input id="searchString" type="text" placeholder="Enter Keywords from title, artist, etc... OR a Spotify or Apple share link" class="form-control">
              <!---->
              <!---->
              <!---->
            </div>
          </div>
          <fieldset class="form-group">
            <legend tabindex="-1" class="bv-no-focus-ring col-form-label pt-0">Dances:</legend>
            <div>
              <div style="border: 1px solid #ced4da;">
                <div role="group" class="b-form-tags form-control h-auto mb-2 mt-2" style="border-style: hidden;"><output role="status" aria-live="off" aria-atomic="true" aria-relevant="additions text" class="sr-only"></output>
                  <div role="status" aria-live="off" aria-atomic="true" class="sr-only"></div>
                  <!---->
                  <div class="dropdown b-dropdown tag-dropdown">
                    <!----><button aria-haspopup="menu" aria-expanded="false" type="button" class="btn dropdown-toggle btn-outline-secondary btn-sm btn-block"><svg viewBox="0 0 16 16" width="1em" height="1em" focusable="false" role="img" aria-label="tag fill" xmlns="http://www.w3.org/2000/svg" fill="currentColor" class="bi-tag-fill b-icon bi">
                        <g>
                          <path d="M2 1a1 1 0 0 0-1 1v4.586a1 1 0 0 0 .293.707l7 7a1 1 0 0 0 1.414 0l4.586-4.586a1 1 0 0 0 0-1.414l-7-7A1 1 0 0 0 6.586 1H2zm4 3.5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"></path>
                        </g>
                      </svg> Choose Dances
                    </button>
                    <ul role="menu" tabindex="-1" class="dropdown-menu">
                      <li role="presentation">
                        <form tabindex="-1" class="b-dropdown-form">
                          <div role="group" class="form-row form-group mb-0"><label for="tag-search-input" class="col-md-auto col-form-label col-form-label-sm">Search Dances</label>
                            <div class="col"><input type="search" autocomplete="off" class="form-control form-control-sm">
                              <!---->
                              <!---->
                              <!---->
                            </div>
                          </div>
                        </form>
                      </li>
                      <li role="presentation">
                        <hr role="separator" aria-orientation="horizontal" class="dropdown-divider">
                      </li>
                      <li role="presentation">
                        <p class="b-dropdown-text">
                          No more dances to choose
                        </p>
                      </li>
                      <div></div>
                    </ul>
                  </div>
                  <!---->
                </div>
                <div class="d-flex justify-content-between w-100 mx-1 mb-2">
                  <div id="dance-connector" role="radiogroup" tabindex="-1" class="mx-3 btn-group-toggle btn-group btn-group-sm bv-no-focus-ring"><label class="btn btn-outline-secondary btn-sm disabled active"><input type="radio" name="dance-connector" disabled="disabled" class="" value="any">Any</label> <label class="btn btn-outline-secondary btn-sm disabled"><input type="radio" name="dance-connector" disabled="disabled" class="" value="all">All</label></div>
                </div>
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
          <fieldset class="form-group">
            <legend tabindex="-1" class="bv-no-focus-ring col-form-label pt-0">Include Tags:</legend>
            <div>
              <div role="group" class="b-form-tags form-control h-auto mb-2" style="border-style: hidden;" id="includeTags"><output role="status" aria-live="off" aria-atomic="true" aria-relevant="additions text" class="sr-only"></output>
                <div role="status" aria-live="off" aria-atomic="true" class="sr-only"></div>
                <!---->
                <div class="dropdown b-dropdown tag-dropdown">
                  <!----><button aria-haspopup="menu" aria-expanded="false" type="button" class="btn dropdown-toggle btn-outline-secondary btn-sm btn-block"><svg viewBox="0 0 16 16" width="1em" height="1em" focusable="false" role="img" aria-label="tag fill" xmlns="http://www.w3.org/2000/svg" fill="currentColor" class="bi-tag-fill b-icon bi">
                      <g>
                        <path d="M2 1a1 1 0 0 0-1 1v4.586a1 1 0 0 0 .293.707l7 7a1 1 0 0 0 1.414 0l4.586-4.586a1 1 0 0 0 0-1.414l-7-7A1 1 0 0 0 6.586 1H2zm4 3.5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"></path>
                      </g>
                    </svg> Choose Tags to Include
                  </button>
                  <ul role="menu" tabindex="-1" class="dropdown-menu">
                    <li role="presentation">
                      <form tabindex="-1" class="b-dropdown-form">
                        <div role="group" class="form-row form-group mb-0"><label for="tag-search-input" class="col-md-auto col-form-label col-form-label-sm">Search Tags</label>
                          <div class="col"><input type="search" autocomplete="off" class="form-control form-control-sm">
                            <!---->
                            <!---->
                            <!---->
                          </div>
                        </div>
                      </form>
                    </li>
                    <li role="presentation">
                      <hr role="separator" aria-orientation="horizontal" class="dropdown-divider">
                    </li>
                    <li role="presentation">
                      <p class="b-dropdown-text">
                        No more tags to choose
                      </p>
                    </li>
                    <li role="presentation">
                      <p class="b-dropdown-text">
                        Start typing to see a list of matching tags
                      </p>
                    </li>
                  </ul>
                </div>
                <!---->
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
          <fieldset class="form-group">
            <legend tabindex="-1" class="bv-no-focus-ring col-form-label pt-0">Exclude Tags:</legend>
            <div>
              <div role="group" class="b-form-tags form-control h-auto mb-2" style="border-style: hidden;" id="excludeTags"><output role="status" aria-live="off" aria-atomic="true" aria-relevant="additions text" class="sr-only"></output>
                <div role="status" aria-live="off" aria-atomic="true" class="sr-only"></div>
                <!---->
                <div class="dropdown b-dropdown tag-dropdown">
                  <!----><button aria-haspopup="menu" aria-expanded="false" type="button" class="btn dropdown-toggle btn-outline-secondary btn-sm btn-block"><svg viewBox="0 0 16 16" width="1em" height="1em" focusable="false" role="img" aria-label="tag fill" xmlns="http://www.w3.org/2000/svg" fill="currentColor" class="bi-tag-fill b-icon bi">
                      <g>
                        <path d="M2 1a1 1 0 0 0-1 1v4.586a1 1 0 0 0 .293.707l7 7a1 1 0 0 0 1.414 0l4.586-4.586a1 1 0 0 0 0-1.414l-7-7A1 1 0 0 0 6.586 1H2zm4 3.5a1.5 1.5 0 1 1-3 0 1.5 1.5 0 0 1 3 0z"></path>
                      </g>
                    </svg> Choose Tags to Exclude
                  </button>
                  <ul role="menu" tabindex="-1" class="dropdown-menu">
                    <li role="presentation">
                      <form tabindex="-1" class="b-dropdown-form">
                        <div role="group" class="form-row form-group mb-0"><label for="tag-search-input" class="col-md-auto col-form-label col-form-label-sm">Search Tags</label>
                          <div class="col"><input type="search" autocomplete="off" class="form-control form-control-sm">
                            <!---->
                            <!---->
                            <!---->
                          </div>
                        </div>
                      </form>
                    </li>
                    <li role="presentation">
                      <hr role="separator" aria-orientation="horizontal" class="dropdown-divider">
                    </li>
                    <li role="presentation">
                      <p class="b-dropdown-text">
                        No more tags to choose
                      </p>
                    </li>
                    <li role="presentation">
                      <p class="b-dropdown-text">
                        Start typing to see a list of matching tags
                      </p>
                    </li>
                  </ul>
                </div>
                <!---->
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </fieldset>
          <div id="tempo-range-group" role="group" class="form-group"><label id="tempo-range-group__BV_label_" for="tempo-range" class="d-block">Tempo range (BPM):</label>
            <div>
              <fieldset id="tempo-range" class="form-group">
                <!---->
                <div>
                  <div class="d-flex"><input id="tempo-min" type="number" min="0" max="400" class="form-control" style="width: 6rem;"> <span class="mx-2 pt-2">to</span> <input id="tempo-max" type="number" min="0" max="400" class="form-control" style="width: 6rem;">
                    <div class="invalid-feedback">
                      Tempos must be between 0 and 400 BPM
                    </div>
                  </div>
                  <!---->
                  <!---->
                  <!---->
                </div>
              </fieldset>
              <!---->
              <!---->
              <!---->
            </div>
          </div>
          <div id="length-range-group" role="group" class="form-group"><label id="length-range-group__BV_label_" for="length-range" class="d-block">Length range (seconds):</label>
            <div>
              <fieldset id="length-range" class="form-group">
                <!---->
                <div>
                  <div class="d-flex"><input id="length-min" type="number" min="0" max="600" class="form-control" style="width: 6rem;"> <span class="mx-2 pt-2">to</span> <input id="length-max" type="number" min="0" max="600" class="form-control" style="width: 6rem;">
                    <div class="invalid-feedback">
                      Length must be between 0 and 600 seconds
                    </div>
                  </div>
                  <!---->
                  <!---->
                  <!---->
                </div>
              </fieldset>
              <!---->
              <!---->
              <!---->
            </div>
          </div>
          <div id="activity-group" role="group" class="form-group"><label id="activity-group__BV_label_" for="activity" class="d-block">By User:</label>
            <div>
              <div class="d-flex"><input id="user" type="text" placeholder="UserName or me" class="mr-2 form-control" style="width: 10rem;"> <select id="activity" class="custom-select">
                  <option value="NT">Don't filter on user activity</option>
                </select></div>
              <!---->
              <!---->
              <!---->
            </div>
          </div>
          <div id="services-group" role="group" class="form-group mx-2 mb-2"><label id="services-group__BV_label_" for="services" class="d-block">Available on:</label>
            <div>
              <div id="services" role="group" tabindex="-1" class="bv-no-focus-ring">
                <div class="custom-control custom-control-inline custom-checkbox"><input type="checkbox" name="services" class="custom-control-input" value="A"><label class="custom-control-label">Amazon</label></div>
                <div class="custom-control custom-control-inline custom-checkbox"><input type="checkbox" name="services" class="custom-control-input" value="I"><label class="custom-control-label">ITunes</label></div>
                <div class="custom-control custom-control-inline custom-checkbox"><input type="checkbox" name="services" class="custom-control-input" value="S"><label class="custom-control-label">Spotify</label></div>
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </div>
          <div id="bonuses-group" role="group" class="form-group mx-2 mb-2"><label id="bonuses-group__BV_label_" for="bonuses" class="d-block">Bonus content:</label>
            <div>
              <div id="bonuses" role="group" tabindex="-1" class="bv-no-focus-ring">
                <div class="custom-control custom-control-inline custom-checkbox"><input type="checkbox" name="bonuses" class="custom-control-input" value="P"><label class="custom-control-label">Not found in any publisher catalog</label></div>
                <div class="custom-control custom-control-inline custom-checkbox"><input type="checkbox" name="bonuses" class="custom-control-input" value="D"><label class="custom-control-label">Not categorized by dance</label></div>
              </div>
              <!---->
              <!----><small tabindex="-1" id="bonuses-group__BV_description_" class="form-text text-muted"><a href="https://music4dance.blog/music4dance-help/subscriptions/">Premium content</a></small>
            </div>
          </div>
          <div id="sort-group" role="group" class="form-group"><label id="sort-group__BV_label_" for="sort" class="d-block">Sort By:</label>
            <div><select id="sort" required="required" aria-required="true" class="custom-select">
                <option value="Title">Title</option>
                <option value="Artist">Artist</option>
                <option value="Tempo">Tempo</option>
                <option value="Length">Length</option>
                <option value="Modified">Last Modified</option>
                <option value="Edited">Last Edited</option>
                <option value="Created">When Added</option>
                <option value="Energy">Energy</option>
                <option value="Mood">Mood</option>
                <option value="Beat">Strength of Beat</option>
                <option value="Comments">Comments</option>
                <option value="">Closest Match</option>
              </select>
              <div id="sort-order" role="radiogroup" tabindex="-1" class="mt-2 bv-no-focus-ring">
                <div class="custom-control custom-control-inline custom-radio"><input type="radio" name="sort-order" class="custom-control-input" value="asc"><label class="custom-control-label">Ascending (A-Z, Slow-Fast, Newest-Oldest,
                    Shortest-Longest)</label></div>
                <div class="custom-control custom-control-inline custom-radio"><input type="radio" name="sort-order" class="custom-control-input" value="desc"><label class="custom-control-label">Descending (Z-A, Fast-Slow, Oldest-Newest,
                    Longest-Shortest)</label></div>
              </div>
              <!---->
              <!---->
              <!---->
            </div>
          </div>
          <div class="d-flex justify-content-between w-100 mx-1 mb-2"><button type="reset" class="btn btn-secondary">Reset</button> <button type="search" class="btn btn-primary">Submit</button></div>
        </form>
        <!---->
      </div>
    </div>
  </div>
  <div id="footer-content">
    <hr>
    <footer>
      <p>
        © 2022 -
        <a href="https://www.music4dance.net">Music4Dance.net</a> -
        <a href="https://www.music4dance.net/home/sitemap">Site Map</a> -
        <a href="https://www.music4dance.net/home/termsofservice">Terms of Service</a>
        -
        <a href="https://www.music4dance.net/home/privacypolicy">Privacy Policy</a>
        - <a href="https://www.music4dance.net/home/credits">Credits</a> -
        <a href="https://music4dance.blog/music4dance-help/song-list/">Help</a>
      </p>
    </footer>
  </div>
  <!---->
</div>
`;
